dask/array/gufunc.py

```diff
@@ -0,0 +1,343 @@
+    core_shapes = merge(output_sizes, *_core_input_shapes)
+
+    loop_input_dimss = [tuple("__loopdim%d__" % d for d in range(max_loopdims - n, max_loopdims)) for n in num_loopdims]
+    loop_input_shapes = [s[:n] for s, n in zip(input_shapes, num_loopdims)]
+
+    input_dimss = [l + c for l, c in zip(loop_input_dimss, core_input_dimss)]
+
+    loop_output_dims = max(loop_input_dimss, key=len) if loop_input_dimss else set()
+    loop_output_shape = broadcast_shapes(*loop_input_shapes)
+
+    ## Prepare function
+    if vectorize:
+        signature = compile_signature(core_input_dimss, core_output_dimss)
+        ufunc = np_vectorize(func, signature=signature)
+    else:
+        # Wrapper that broadcasts only loop dims (as opposed to numpy.broadcast_arrays)
+        def ufunc(*args, **kwargs):
+            input_shapes = (a.shape for a in args)
+            loop_input_blockshapes = (s[:n] for s, n in zip(input_shapes, num_loopdims))
+            loop_output_blockshape = broadcast_shapes(*loop_input_blockshapes)
+            shapes = (loop_output_blockshape + a.shape[n:] for a, n in zip(args, num_loopdims))
+            bargs = (np_broadcast_to(a, s) for a, s in zip(args, shapes))
+            return func(*bargs, **kwargs)
+
+    ## Apply function - use atop here
+    arginds = list(concat(zip(args, input_dimss)))
+
+    ### Treat direct output
+    if nout is None:
+        core_output_dimss = [core_output_dimss]
+        output_dtypes = [output_dtypes]
+
+    ### Use existing `atop` but only with loopdims to enforce
+    ### concatenation, for coredims, that appear also at the output
+    ### Modifying `atop` could improve things here.
+    tmp = atop(ufunc, loop_output_dims, *arginds,
+               dtype=int,  # Only dummy dtype, anyone will do
+               concatenate=concatenate,
+               **kwargs)
+
+    ## Prepare output shapes
+    assert loop_output_shape == tmp.shape
+    loop_output_chunks = tmp.chunks
+    dsk = tmp.__dask_graph__()
+    keys = list(flatten(tmp.__dask_keys__()))
+    _anykey = keys[0]
+    name, token = _anykey[0].split('-')
+
+    ## Split output
+    leaf_arrs = []
+    for i, cod, odt in zip(count(0), core_output_dimss, output_dtypes):
+        core_output_shape = tuple(core_shapes[d] for d in cod)
+        core_chunkinds = len(cod) * (0,)
+        output_shape = loop_output_shape + core_output_shape
+        output_chunks = loop_output_chunks + core_output_shape
+        leaf_name = "%s_%d-%s" % (name, i, token)
+        leaf_dsk = {(leaf_name,) + key[1:] + core_chunkinds: ((getitem, key, i) if nout else key) for key in keys}
+        leaf_arr = Array(sharedict.merge((leaf_name, leaf_dsk), dsk),
+                         leaf_name,
+                         chunks=output_chunks,
+                         shape=output_shape,
+                         dtype=odt)
+        leaf_arrs.append(leaf_arr)
+
+    return leaf_arrs if nout else leaf_arrs[0]
+
+
+@curry
```

Comment:
Using curry can hide odd sorts of errors.  If we know how this is likely to be used then it might be cleaner to make a function that returns another function explicitly.

Reply:
I have removed it for now. Check for correctness of passed `**kwargs` is still postponed until `apply_gufunc` is called, i.e. very late. Depending on the other discussion about dtype inference, we can finalize the behavior here eventually